tests/UpgradeRule/PHP/Visitor/PropertyWarningsVisitorTest.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace SilverStripe\Upgrader\Tests\UpgradeRule\PHP\Visitor;
+
+use SilverStripe\Upgrader\UpgradeRule\PHP\Visitor\PropertyWarningsVisitor;
+use SilverStripe\Upgrader\Util\ApiChangeWarningSpec;
+
+class PropertyWarningsVisitorTest extends BaseVisitorTest
+{
+    public function testInstancePropDefinitionWithoutClass()
+    {
+
+        $input = <<<PHP
+<?php
+
+namespace MyNamespace;
+
+use SomeNamespaced\NamespacedClass;
+
+class MyClass
+{
+    protected \$removedInstanceProp = true;
+}
+
+class MyOtherClass
+{
+    function useProp()
+    {
+        \$foo = new MyClass();
+        \$foo->removedInstanceProp;
+    }
+    
+    function noMatch()
+    {
+        \$bar = new MyOtherClass();
+        \$bar->removedInstanceProp;
+    }
+}
+PHP;
+
+        $visitor = new PropertyWarningsVisitor([
+            (new ApiChangeWarningSpec('removedInstanceProp', 'Test instance prop'))
+        ], $this->getMockFile($input));
+
+        $this->traverseWithVisitor($input, $visitor);
+
+        $warnings = $visitor->getWarnings();
+        $this->assertCount(3, $warnings);
+
+        $this->assertContains('Test instance prop', $warnings[0]->getMessage());
+        $this->assertContains('protected $removedInstanceProp', $this->getLineForWarning($input, $warnings[0]));
+
+        $this->assertContains('Test instance prop', $warnings[1]->getMessage());
+        $this->assertContains('$foo->removedInstanceProp', $this->getLineForWarning($input, $warnings[1]));
+
+        $this->assertContains('Test instance prop', $warnings[2]->getMessage());
+        $this->assertContains('$bar->removedInstanceProp', $this->getLineForWarning($input, $warnings[2]));
+    }
+
+    public function testInstancePropDefinitionWithClass()
+    {
+
+        $input = <<<PHP
+<?php
+
+namespace MyNamespace;
+
+use SomeNamespaced\NamespacedClass;
+
+class MyClass
+{
+    protected \$removedInstanceProp = true;
+}
+
+class MyOtherClass
+{
+    function useProp()
+    {
+        \$foo = new MyClass();
+        \$foo->removedInstanceProp;
+    }
+    
+    function noMatch()
+    {
+        \$bar = new MyOtherClass();
+        \$bar->removedInstanceProp;
+    }
+}
+PHP;
+
+        $visitor = new PropertyWarningsVisitor([
+            (new ApiChangeWarningSpec('MyNamespace\\MyClass->removedInstanceProp', 'Test instance prop'))
```

Comment:
What about subclasses of MyNamespace\MyClass? You should add a test for that.

Reply:
This may require upgrader to build a manifest of the entire class hierarchy; I.e. context not available within the specific file being parsed.

Reply:
Yeah, definitely an SSViewer-esque slippery slope ;) I think for now, I think a "class-qualified method/prop type inference" is very limited - as you can see on the related framework PR, I've left most instance method/prop without a class qualifier for this reason. I'm keen to investigate how feasible https://github.com/phpstan/phpstan would be instead of this context hackery before taking this approach much further. For now, I think it covers the 80/20 rule pretty well, maybe with a handful of preventable false positives.